python/ray/serve/_private/logging_utils.py

```diff
@@ -74,12 +75,27 @@ def format(self, record: logging.LogRecord) -> str:
             ]
         if SERVE_LOG_ROUTE in record.__dict__:
             record_format[SERVE_LOG_ROUTE] = SERVE_LOG_RECORD_FORMAT[SERVE_LOG_ROUTE]
+
         if SERVE_LOG_APPLICATION in record.__dict__:
             record_format[SERVE_LOG_APPLICATION] = SERVE_LOG_RECORD_FORMAT[
                 SERVE_LOG_APPLICATION
             ]
 
-        record_format[SERVE_LOG_MESSAGE] = SERVE_LOG_RECORD_FORMAT[SERVE_LOG_MESSAGE]
+        if SERVE_LOG_MESSAGE in record.__dict__:
+            record_format[SERVE_LOG_MESSAGE] = (
+                SERVE_LOG_RECORD_FORMAT[SERVE_LOG_MESSAGE] % record.__dict__
+            )
+
+        if SERVE_LOG_EXTRA_FIELDS in record.__dict__:
+            if not isinstance(record.__dict__[SERVE_LOG_EXTRA_FIELDS], dict):
+                raise ValueError(
+                    f"Expected a dictionary passing into {SERVE_LOG_EXTRA_FIELDS}, "
+                    f"but got {type(record.__dict__[SERVE_LOG_EXTRA_FIELDS])}"
+                )
+            for k, v in record.__dict__[SERVE_LOG_EXTRA_FIELDS].items():
+                if k in record_format:
+                    raise KeyError(f"Found duplicated key in the log record: {k}")
```

Comment:
Nitpick (non-blocker): would be nice to add a unit test for this and when `record.__dict__[SERVE_LOG_EXTRA_FIELDS]` is not a dictionary 🙂

Reply:
done